File: frontend/src/services/contributionScoreService.ts
import axiosInstance from './axiosInstance';
import { ContributionScoreResponse } from '../types/contribution';

class ContributionScoreService {
  // Get all contribution scores for a project
  async getScoresByProject(projectId: number) {
    try {
      const response = await axiosInstance.get(`/api/contribution-scores/projects/${projectId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching contribution scores:', error);
      throw error;
    }
  }

  // Get all contribution scores for a group
  async getScoresByGroup(groupId: number) {
    try {
      const response = await axiosInstance.get(`/api/contribution-scores/groups/${groupId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching group contribution scores:', error);
      throw error;
    }
  }

  // Get contribution score for a specific user in a project
  async getScoreByUserAndProject(projectId: number, userId: number) {
    try {
      const response = await axiosInstance.get(`/api/contribution-scores/projects/${projectId}/users/${userId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching user contribution score:', error);
      throw error;
    }
  }

  // Adjust a user's contribution score (for instructors/admins)
  async adjustScore(scoreId: number, adjustedScore: number, adjustmentReason: string) {
    try {
      const response = await axiosInstance.put(`/api/contribution-scores/${scoreId}/adjust`, {
        adjustedScore,
        adjustmentReason
      });
      return response.data;
    } catch (error) {
      console.error('Error adjusting contribution score:', error);
      throw error;
    }
  }

  // Finalize all contribution scores for a project
  async finalizeScores(projectId: number) {
    try {
      const response = await axiosInstance.put(`/api/contribution-scores/projects/${projectId}/finalize`);
      return response.data;
    } catch (error) {
      console.error('Error finalizing contribution scores:', error);
      throw error;
    }
  }
}

export default new ContributionScoreService();

File: backend/src/main/java/com/itss/projectmanagement/service/ContributionScoreService.java
package com.itss.projectmanagement.service;

import com.itss.projectmanagement.dto.response.contribution.ContributionScoreResponse;
import com.itss.projectmanagement.entity.ContributionScore;
import com.itss.projectmanagement.entity.Project;
import com.itss.projectmanagement.entity.User;

import java.util.List;

public interface ContributionScoreService {
    /**
     * Calculate contribution score for a user in a project
     * @param user The user
     * @param project The project
     * @return The calculated contribution score
     */
    ContributionScore calculateScore(User user, Project project);
    
    /**
     * Calculate contribution scores for all users in a project
     * @param project The project
     * @return List of calculated contribution scores
     */
    List<ContributionScore> calculateScoresForProject(Project project);
    
    /**
     * Get contribution score for a specific user in a project
     * @param user The user
     * @param project The project
     * @return The contribution score response DTO
     */
    ContributionScoreResponse getScoreByUserAndProject(User user, Project project);
    
    /**
     * Get contribution scores for all users in a project
     * Always recalculates the latest scores
     * @param project The project
     * @return List of contribution score response DTOs
     */
    List<ContributionScoreResponse> getScoresByProject(Project project);
    
    /**
     * Adjust a user's contribution score (by instructor)
     * @param id The contribution score ID
     * @param adjustedScore The adjusted score
     * @param adjustmentReason The reason for adjustment
     * @return The updated contribution score response DTO
     */
    ContributionScoreResponse adjustScore(Long id, Double adjustedScore, String adjustmentReason);
    
    /**
     * Finalize contribution scores for a project
     * @param projectId The project ID
     * @return List of finalized contribution score response DTOs
     */
    List<ContributionScoreResponse> finalizeScores(Long projectId);
}